bin/target.ml

```diff
@@ -23,32 +23,37 @@ let target_hint (_setup : Dune_rules.Main.build_system) path =
   assert (Path.is_managed path);
   let open Memo.O in
   let sub_dir = Option.value ~default:path (Path.parent path) in
-  (* CR-someday amokhov: There are two issues with the code below.
-
-     (1) We first get *all* targets but then filter out only those that are
-     defined in the [sub_dir]. It would be better to just get the targets for
-     the [sub_dir] directly (the API supports this).
-
-     (2) We currently provide the same hint for all targets. It would be nice to
-     indicate whether a hint corresponds to a file or to a directory target. *)
-  let+ candidates = Load_rules.all_direct_targets () >>| Path.Build.Map.keys in
-  let candidates =
-    if Path.is_in_build_dir path then List.map ~f:Path.build candidates
-    else
-      List.map candidates ~f:(fun path ->
-          match Path.Build.extract_build_context path with
-          | None -> Path.build path
-          | Some (_, path) -> Path.source path)
+  let root =
+    match sub_dir with
+    | External _ -> (* checked above *) assert false
```

Comment:
By "checked above", do you mean `assert (Path.is_managed path)`? If yes, how about we just remove that assert, and replace this one with a code error?